### ISSUE TITLE
Testing issues with chalk@2

### ISSUE THREAD

**Issue by coreyfarrell** [yargs/cliui#59]
I've tested cliui against newer dependencies, upgrading chalk to 2.x causes a test failure:
```
  1) cliui
       layoutDSL
         ignores ansi escape codes when measuring padding:

      AssertionError: expected [ Array(9) ] to deeply equal [ Array(5) ]
      + expected - actual

       [
         "  |"
      -  "  "
         "  __|   __|  |   |   _ \\"
      -  "  "
         "  |    |     |   |   __/"
      -  "  "
         " \\__| _|    \\__,_| \\___|"
      -  " "
         "                         "
       ]
```

I've tested the following script:
```js
const chalk = require('chalk');
console.log(JSON.stringify(chalk.blue('line 1\nline 2').split('\n'), null, 4));
```
Result from `chalk@^1.1.2`:
```json
[
    "\u001b[34mline 1",
    "line 2\u001b[39m"
]
```

Result from `chalk@^2.4.2`:
```json
[
    "\u001b[34mline 1\u001b[39m",
    "\u001b[34mline 2\u001b[39m"
]
```

So the big difference is that `chalk@2` adds ansi codes to the start/end of each line.  Upgrading `wrap-ansi` to the latest didn't fix the testing errors, it actually caused more errors though I'm unsure if this is due to an issue with `wrap-ansi` or with `cliui`.

**Comment by bcoe** [comment]
@coreyfarrell is introducing ansi codes to the beginning and end of each line an intentional change in `chalk@2`? It seems pretty innocent, I wouldn't think it would break anything.

I'm more concerned about the updated version of `wrap-ansi` causing test failures.

**Comment by coreyfarrell** [comment]
chalk/chalk#92 looks intentional to deal with an OSX terminal bug.

Both issues need to be fixed so cliui can update everything, but `wrap-ansi` is a direct dependency which means that `cliui` controls which version is installed / prevents the bug from manifesting.  `chalk` is a devDependency so `cliui` has to deal with whatever it's given.

I've opened chalk/wrap-ansi#32 to add a failing test case.  That would fix `padding -> preserves leading whitespace as padding` and partially fix `layoutDSL -> ignores ansi escape codes when measuring padding`.  I believe the second test case should be adjusted to expect the last line to be empty, the failing to strip trailing spaces inside color codes was a bug in `wrap-ansi@2.1.0`.

I have a feeling that the wrap-ansi fix will correct both issues.

**Comment by coreyfarrell** [comment]
@bcoe I was correct.  I tested cliui with all dependencies updated then applied chalk/wrap-ansi#33, no test failures.  Once that is merged / released I'll submit a PR to update cliui.  It will cause node.js 6 to be required but that should be fine since yargs already requires the same.

**Comment by bcoe** [comment]
@coreyfarrell sounds good, will happily accept this patch.

**Comment by coreyfarrell** [comment]
@bcoe my patch yesterday accidentally updated string-width to 4.1.0 which requires node.js 8, now fixed to only upgrade to string-width 3.1.0.

Bonus: once this is merged and yargs is updated `npm i yargs` will not install any duplicate packages (only one version of each dependency will be used).
